operator/v1/types_dns.go

```diff
@@ -130,6 +131,70 @@ type Server struct {
 	ForwardPlugin ForwardPlugin `json:"forwardPlugin"`
 }
 
+// DNSTransport indicates what type of connection should be used
+// +kubebuilder:validation:Enum=TLS;Cleartext
+type DNSTransport string
+
+const (
+	// TLSTransport indicates that TLS should be used for the connection.
+	TLSTransport DNSTransport = "TLS"
+
+	// CleartextTransport indicates that no encryption should be used for
+	// the connection.
+	CleartextTransport DNSTransport = "Cleartext"
+)
+
+// DNSTransportConfig groups related configuration parameters used for configuring
+// forwarding to upstream resolvers that support DNS-over-TLS.
+// +union
+type DNSTransportConfig struct {
+	// transport allows cluster administrators to opt-in to using a DNS-over-TLS
+	// connection between cluster DNS and an upstream resolver(s). Configuring
+	// TLS as the transport at this level without configuring a CABundle will
+	// result in the system certificates being used to verify the serving
+	// certificate of the upstream resolver(s).
+	//
+	// Possible values:
+	// "" (empty) - This means no explicit choice has been made and the platform chooses the default which is subject
+	// to change over time. The current default is "Cleartext".
+	// "Cleartext" - Cluster admin specified cleartext option. This results in the same functionality
+	// as an empty value but may be useful when a cluster admin wants to be more explicit about the transport,
+	// or wants to switch from "TLS" to "Cleartext" explicitly.
+	// "TLS" - This indicates that DNS queries should be sent over a TLS connection. If Transport is set to TLS,
+	// you MUST also set ServerName.
+	//
+	// +optional
+	// +unionDiscriminator
+	Transport DNSTransport `json:"transport,omitempty"`
+
+	TLS *DNSOverTLSConfig `json:"tls,omitempty"`
+}
+
+// DNSOverTLSConfig describes optional DNSTransportConfig fields that should be captured.
+type DNSOverTLSConfig struct {
+	// serverName is the upstream server to connect to for DNS resolution. This is required when Transport is set to
+	// "TLS". ServerName will be validated against the DNS naming conventions in RFC 1123 and should match the TLS
+	// certificate installed in the upstream resolver(s).
+	//
+	// + ---
+	// + Inspired by the DNS1123 patterns in Kubernetes: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L178
```

Comment:
Please use a permalink.  (Good idea to add the citation though!)

Reply:
Done!